pkg/maintainer/spv/redemptions.go

```diff
@@ -0,0 +1,312 @@
+			)
+			continue
+		}
+
+		walletTransactions, err := btcChain.GetTransactionsForPublicKeyHash(
+			walletPublicKeyHash,
+			transactionLimit,
+		)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to get transactions for wallet: [%v]",
+				err,
+			)
+		}
+
+		for _, transaction := range walletTransactions {
+			isUnproven, err :=
+				isUnprovenRedemptionTransaction(
+					transaction,
+					walletPublicKeyHash,
+					btcChain,
+					spvChain,
+				)
+			if err != nil {
+				return nil, fmt.Errorf(
+					"failed to check if transaction is an unproven redemption "+
+						"transaction: [%v]",
+					err,
+				)
+			}
+
+			if isUnproven {
+				unprovenRedemptionTransactions = append(
+					unprovenRedemptionTransactions,
+					transaction,
+				)
+			}
+		}
+	}
+
+	return unprovenRedemptionTransactions, nil
+}
+
+func isUnprovenRedemptionTransaction(
+	transaction *bitcoin.Transaction,
+	walletPublicKeyHash [20]byte,
+	btcChain bitcoin.Chain,
+	spvChain Chain,
+) (bool, error) {
+	// If the transaction does not have exactly one input, it cannot be a
+	// redemption transaction.
+	if len(transaction.Inputs) != 1 {
+		return false, nil
+	}
+
+	singleInput := transaction.Inputs[0]
+
+	// Check whether the single input is the current wallet main UTXO.
+	isMainUtxo, err := isInputCurrentWalletsMainUTXO(
+		singleInput.Outpoint.TransactionHash,
+		singleInput.Outpoint.OutputIndex,
+		walletPublicKeyHash,
+		btcChain,
+		spvChain,
+	)
+	if err != nil {
+		return false, fmt.Errorf(
+			"failed to check if input is the main UTXO",
+		)
+	}
+
+	// If the single input is not the current main UTXO of the wallet, the
+	// transaction is either a redemption transaction that is already
+	// proven or it's not a redemption transaction at all.
+	if !isMainUtxo {
+		return false, nil
+	}
+
+	changeFound := false
+
+	// Look at the transaction's outputs. All the outputs must be pending
+	// redemption requests, except for one optional change output.
+	for _, output := range transaction.Outputs {
+		// First, check if the given output is a change (if it wasn't
+		// found yet).
+		if !changeFound {
+			isChange, err := isWalletChangeOutput(walletPublicKeyHash, output)
+			if err != nil {
+				return false, fmt.Errorf(
+					"failed to check if output is wallet change: [%v]",
+					err,
+				)
+			}
+
+			if isChange {
+				changeFound = true
+				continue
+			}
+		}
+
+		// If the given output is not a change, it must be a pending redemption
+		// request.
+		_, err := spvChain.GetPendingRedemptionRequest(
+			walletPublicKeyHash,
+			output.PublicKeyScript,
+		)
+		if err != nil {
+			if errors.Is(err, tbtc.ErrPendingRedemptionRequestNotFound) {
+				// This output is neither a change nor a pending request.
+				// That means this is not a redemption transaction.
+				return false, nil
+			} else {
+				return false, fmt.Errorf(
+					"failed to get pending redemption request: [%w]",
+					err,
+				)
+			}
+		}
+	}
+
+	return true, nil
```

Comment:
I would add a check that at least one output represented a redemption, i.e. the change output was not the only output. We check this condition in the Bridge.

Reply:
Good point! See c372c0ce